### PR TITLE
[Core] Add ~/.local/bin to make `which ray` work if ray is installed in `~/.local`

### PR DESCRIPTION
**Pull request by Michaelvll**
To reproduce:
1. `docker build -t mydockerrepo/image-name:v1 .`; `docker push mydockerrepo/image-name:v1`
```Dockerfile
FROM berkeleyskypilot/skypilot-k8s

# Use a RUN command to modify PATH and write to a temporary file
RUN echo "export PATH=$(echo $PATH | sed -e 's|/home/sky/.local/bin:||g' -e 's|:/home/sky/.local/bin||g')" > ~/.bashrc
```
2. `sky launch -c test --image-id docker:mydockerrepo/image-name:v1 echo hi`

Another Dockerfile to test:
``````Dockerfile
FROM berkeleyskypilot/skypilot-k8s

# Use a RUN command to modify PATH and write to a temporary file
RUN echo "export PATH=$(echo $PATH | sed -e 's|/home/sky/.local/bin:||g' -e 's|:/home/sky/.local/bin||g')" > ~/.bashrc

RUN pip uninstall ray -y
``````

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] Reproducible code above.
- [x] All smoke tests: `pytest tests/test_smoke.py` (except TPU tests #3425)
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [x] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/skylet/constants.py [hunk 1]
```diff
@@ -121,12 +123,18 @@
     # latest ray port 6380, but those existing cluster launched before #1790
     # that has ray cluster on the default port 6379 will be upgraded and
     # restarted.
+    'echo PATH=$PATH; '
```
Comment (concretevitamin) [inline sky/skylet/constants.py:126]: Is this for debugging? Worth a comment.

File: sky/skylet/constants.py [hunk 2]
```diff
@@ -121,12 +123,18 @@
...
     f'{SKY_PIP_CMD} list | grep "ray " | '
     f'grep {SKY_REMOTE_RAY_VERSION} 2>&1 > /dev/null '
     f'|| {RAY_STATUS} || '
-    f'{SKY_PIP_CMD} install --exists-action w -U ray[default]=={SKY_REMOTE_RAY_VERSION};'  # pylint: disable=line-too-long
+    f'{SKY_PIP_CMD} install --exists-action w -U ray[default]=={SKY_REMOTE_RAY_VERSION}; '  # pylint: disable=line-too-long
+    # Add missing PATH to make sure ray is in the PATH, when the
+    # previous ray installation happens in user's `~/.local` directory.
+    # ~/.local/bin is added to the end of PATH to avoid conflicts with ray just
+    # installed in the conda environment.
+    'export PATH=$PATH:$HOME/.local/bin; '
```
Comment (concretevitamin) [inline sky/skylet/constants.py:135]: Do I understand correctly:
- The issue is in some envs, our own Ray is by default installed into ~/.local/bin 
- In that env, this path is not in $PATH by default
- So, previously we touch the SKY_RAY_PATH_FILE file, which is empty
- Bug is, later on when we do `our_ray start`, `our_ray` is basically empty string?

**Review by concretevitamin — APPROVED** [review]
LGTM with a question, thanks @Michaelvll.

File: sky/skylet/constants.py (hunk 1 above)
Comment (Michaelvll) [reply sky/skylet/constants.py]: Yes, just added the comment.

File: sky/skylet/constants.py (hunk 2 above)
Comment (Michaelvll) [reply sky/skylet/constants.py:140]: Yes, this is exactly the issue. Tried to elaborate the comments a bit. PTAL : )

**Review by concretevitamin — APPROVED** [review]

**Pull request merged**